File: LeetCode/Java/115-dynamic-programming-不同的子序列.java
/*
 * @lc app=leetcode.cn id=115 lang=java
 *
 * [115] 不同的子序列
 */

// @lc code=start
class Solution {
    public int numDistinct(String s, String t) {
        int m = t.length();
        int n = s.length();
        int[][] dp = new int[m + 1][n + 1];
        for (int i = 0; i <= n; i++)
            dp[0][i] = 1;
        for (int i = 1; i <= m; i++)
            for (int j = 1; j <= n; j++) {
                if (t.charAt(i - 1) == s.charAt(j - 1))
                    dp[i][j] = dp[i][j - 1] + dp[i - 1][j - 1];
                else
                    dp[i][j] = dp[i][j - 1];
            }
        return dp[m][n];
    }
}
// @lc code=end

File: LeetCode/C/115-dynamic-programming-不同的子序列.cpp
/*
 * @lc app=leetcode.cn id=115 lang=cpp
 *
 * [115] 不同的子序列
 */
#include "main.h"

// @lc code=start
class Solution {
  public:
    int numDistinct(string s, string t) {
        vector<long long> dp(t.size() + 1, 0);
        dp[0] = 1;
        for (int i = 0; i < s.size(); i++)
            for (int j = t.size(); j > 0; j--)
                if (s[i] == t[j - 1])
                    dp[j] += dp[j - 1];
        return (int)dp.back();
    }
};
// @lc code=end
